examples/larger-examples/datalog/subtyping.flix

```diff
@@ -0,0 +1,301 @@
+        let cs    = inject p.classes             into Class;
+        let is    = inject p.interfaces          into Interface;
+        let impls = inject p.classImplements     into ClassImplements;
+        let ce    = inject p.classExtends        into ClassExtends;
+        let ies   = inject p.interfaceExtends    into InterfaceExtends;
+        let fs    = inject p.fields              into Field;
+        let ms    = inject p.classMethods        into ClassMethod;
+        let iss   = inject p.interfaceSignatures into InterfaceSignature;
+        let rules = #{
+            Subtype(x, x)       :- Class(x).
+            Subtype(x, x)       :- Interface(x).
+            Subtype(sub, super) :- Class(sub), Interface(super), ClassImplements(sub, super).
+            Subtype(sub, super) :- Class(sub), Class(super), ClassExtends(sub, super).
+            Subtype(sub, super) :- Interface(sub), Interface(super), InterfaceExtends(sub, super).
+            Subtype(sub, super) :- Subtype(sub, a), Subtype(a, super).
+
+            AvailableField(class, field)        :- Field(t, field), Subtype(class, t).
+            AvailableMethod(class, method)      :- ClassMethod(t, method), Subtype(class, t).
+            NonImplementedMethod(class, method) :- InterfaceSignature(t, method), Class(class), Subtype(class, t), not AvailableMethod(class, method).
+        };
+        #(Subtype, AvailableField, AvailableMethod, NonImplementedMethod) -> rules <+> cs <+> is <+> ce <+> impls <+> ies <+> fs <+> ms <+> iss
+
+    pub def hasField(p: Program, field: String): Vector[String] =
+        query classHierarchyAnalysis(p) select tpe from AvailableField(tpe, field)
+
+    pub def hasMethod(p: Program, method: String): Vector[String] =
+        query classHierarchyAnalysis(p) select tpe from AvailableMethod(tpe, method)
+
+    pub def nonImplementedMethods(p: Program): Vector[(String, String)] =
+        query classHierarchyAnalysis(p) select (tpe, method) from NonImplementedMethod(tpe, method)
+
+    pub def isSubtype(p: Program, subtype: String, possibleSupertype: String): Bool =
+        let res = query classHierarchyAnalysis(p) select subtype from Subtype(subtype, possibleSupertype);
+        not Vector.isEmpty(res)
+
+    pub def illegalFinalClasses(p: Program, finalClasses: Vector[String]): Vector[String] =
+        let fcs  = inject finalClasses into FinalClass;
+        let rule = #{
+            ExtendsFinalClass(sub, super)  :- FinalClass(super), Subtype(sub, super), if sub != super.
+        };
+        query fcs, rule, classHierarchyAnalysis(p) select tpe from ExtendsFinalClass(tpe, _)
+
+    pub def illegalNonThrowable(p: Program, throws: Vector[String]): Vector[String] =
+        let ths = inject throws into Throws;
+        let rule = #{
+            IllegalThrowable(t) :- Throws(t), not Subtype(t, "java.lang.Throwable").
+        };
+        query ths, rule, classHierarchyAnalysis(p) select tpe from IllegalThrowable(tpe)
+
+
```

Comment:
I have an idea for another extension, but maybe its best to wait until after the above refactorings.

But if we extended ConcreteMethod to ConcreteMethod(class, method, source), then we could check that every class that specifies an equals method also remembers to implement a hashCode method.

Reply:
The class on which `method` is implemented?

Reply:
I think you got it right.